### DISCUSSION TITLE
Multiple forms on one page, with different defaultValues

### DISCUSSION THREAD

**Discussion by h0jeZvgoxFepBQ2C** [react-hook-form/react-hook-form#1763] [General] [answered]
Is there any better solutation than this?

```javascript
  const { register, handleSubmit, watch, errors, setError, control } = useForm({
    defaultValues: object1
  })

  const { register2, handleSubmit2, watch2, errors2, setError2, control2 } = useForm({
    defaultValues: object2
  })
```

**Comment by bluebill1049** [comment, marked as answer]
here is a video on building multiple steps forms: https://www.youtube.com/watch?v=CeAkxVwsyMU

**Reply by h0jeZvgoxFepBQ2C** [reply]
Thanks, but I was looking for help when both forms are visible at the same time on the same page.

**Reply by bluebill1049** [reply]
then build two components with individual `useForm`

**Reply by h0jeZvgoxFepBQ2C** [reply]
Like I have shown above?

**Reply by bluebill1049** [reply]
nope, build two separate components, so you don't have to do 1 and 2.

**Reply by h0jeZvgoxFepBQ2C** [reply]
Ah yeah, this could also work 🙈  thanks!

**Reply by Ck088** [reply]
> nope, build two separate components, so you don't have to do 1 and 2.

Yes created but how can we handle submit with one submit button in parent component?

**Reply by bluebill1049** [reply]
so you want to do which form is submitted?

**Reply by Ck088** [reply]
No, I want to handleSubmit from single button click in both child form components.

**Reply by bluebill1049** [reply]
i think you want to use `shouldUnregister: false` and still keep in the same form, so when it submits, it will contains both section's data.

**Reply by irekrog** [reply]
https://stackoverflow.com/questions/60276510/how-to-make-react-hook-form-work-with-multiple-forms-in-one-page @bluebill1049 look at the most upvoted answer and I dont know is it anti-pattern have two useForm in one component?

**Reply by bluebill1049** [reply]
> https://stackoverflow.com/questions/60276510/how-to-make-react-hook-form-work-with-multiple-forms-in-one-page @bluebill1049 look at the most upvoted answer and I dont know is it anti-pattern have two useForm in one component?

you can have as many as you want, it's valid HTML :) It's valid, but you want to keep a large form which divided into section (which involve those section gets unmounted) you can consider to use `shouldUnregister:false`, but personally i prefer separate `form` and collect data in a state lib.

**Reply by KingJune28** [reply]
Please @bluebill1049   How can I effectively use React Hook Form to manage  multiple forms, ensuring that each form's state and validation logic is isolated but still part of a cohesive user interface?



I'm working on a project that requires managing several forms within a single React application. Each form represents different data models, such as incident reports, victim details, damage assessments, and event logs. I need a way to dynamically render these forms based on user interaction, ensuring that each form captures and validates its respective data accurately. The challenge is to maintain a clean and manageable codebase while handling state, validation, and submissions for multiple forms. I'm considering using React Hook Form due to its performance.

**Reply by KingJune28** [reply]
you said you prefer separate form and collect data in a state lib.  @bluebill1049  can you show us some example ?

**Comment by Sristi27** [comment]
How to build two different forms in the same component?

**Reply by Moshyfawn** [reply]
Either divide them into their separate components or call multiple `useForm` hooks and assign different names to form methods like
```js
const { register: registerFirstFormField } = useForm()
const { register: registerSecondFormField } = useForm()
```

Keep in mind, that depending on the complexity of your form, the latter solution can be a performance hit.

I'd go with one `useForm` hook per component

**Reply by Sristi27** [reply]
Will this work if both the forms have different input fields?

**Reply by Moshyfawn** [reply]
Sure, provided, you use the corresponding form instance and it's form methods

```jsx
const MultipleForms = () => {
  const {
    register: registerFirstField,
    handleSubmit: handleFirstFormSubmit
  } = useForm()
  const {
    register: registerSecondField,
    handleSubmit: handleSecondFormSubmit
  } = useForm()

  return (
    <div>
      <form onSubmit={handleFirstFormSubmit(onFirstFormSubmit)}>
        <input {...registerFirstField('firstFormFieldName')} />
      </form>
      <form onSubmit={handleSecondFormSubmit(onSecondFormSubmit)}>
        <input {...registerSecondField('secondFormFieldName')} />
      </form>
    </div>
  )
}
```

[A live CSB example](https://codesandbox.io/s/fervent-morning-w800oq?file=%2Fsrc%2FApp.tsx)

**Reply by Sargnec** [reply]
Is there a way to give 2 different register to 1 input?

**Reply by Moshyfawn** [reply]
> Is there a way to give 2 different register to 1 input? 

No, you'd simply override the event handlers `register` method returns, unless you call the method before your `return` statement and assign the returned props manually. But I don't see why you'd actually need to do this

**Comment by mani-cafs** [comment]
antd 2 different forms how to use useForm in reactjs
